Document default pipeline layout behavior.

BGR360 opened the pull request:
**Connections**
Fixes #6254.

**Description**
The behavior of setting `{Render,Compute}PipelineDescriptor::layout` to `None` was not documented. It is now documented, with links to the WebGPU spec providing more information.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo doc -p wgpu --no-deps --lib`

BGR360 committed:
Document default pipeline layout behavior.

Fixes #6254.

BGR360 committed:
Fix typo

Wumpf commented on wgpu/src/api/compute_pipeline.rs:
```diff
@@ -20,6 +20,13 @@ super::impl_partialeq_eq_hash!(ComputePipeline);
+    /// then bind groups created with the returned `BindGroupLayout` can only be used with this
+    /// pipeline.
+    ///
+    /// # Panics
+    /// - There is no bind group layout at the given `index`.
```
[not actionable, mostly note to self]

urgh this _actually_ panics. 
For future reference: This should be a validation error and return an invalid object, see https://developer.mozilla.org/en-US/docs/Web/API/GPUComputePipeline/getBindGroupLayout#validation

BGR360 replied:
> this *actually* panics

By this, do you mean that there's a bug here? I.e. `wgpu` should be catching a validation error and doing the proper thing with it, but it's not?

Or do you just mean to say "ah yes, I suppose this *does* panic if the user doesn't set any error scopes."

Wumpf replied:
Looked closer again to be sure: Today it panics on wgpu-core which is a simply a bug. Turns out this also got filed already (didn't check before).. twice in fact. 
* https://github.com/gfx-rs/wgpu/issues/4167
* https://github.com/gfx-rs/wgpu/issues/6095

...closing one of them as dupl

Considering that, I think the doc here should say that this is a validation error and then add a note that there is a known bug that this panics

BGR360 replied:
Just FYI, I actually didn't go and research for myself whether it panics. I just made an assumption and hoped that my code reviewer would correct me.

Looking at the implementation, it looks like it does the proper thing on WebGPU, but not on wgpu_core. Given that, I don't think we should put anything in the public docs until we fix wgpu_core to raise this as a validation error.

BGR360 replied:
Oops, I responded before I refreshed and saw your reply. The known bug option sounds fine too.

BGR360 replied:
Okay actually I changed my mind, I'm going to omit any documentation about panics or errors until webgpu and wgpu-core have the same behavior.

Wumpf replied:
fine by me! Didn't want to bog down this otherwise fine doc addition :)
Bit of a wild west situation this particular thing at this point T_T

Wumpf commented on wgpu/src/api/compute_pipeline.rs:
```diff
@@ -48,6 +55,25 @@ pub struct ComputePipelineDescriptor<'a> {
     /// The layout of bind groups for this pipeline.
+    ///
+    /// If this is set, then `wgpu` will validate that the layout matches what the shader module(s)
+    /// expect, otherwise [`Device::create_compute_pipeline`] will panic.
```
```suggestion
    /// expect, otherwise [`Device::create_compute_pipeline`] will cause a validation error.
```

Should only panic with the default error handler.
(see also https://github.com/gfx-rs/wgpu/issues/3767)

Wumpf commented on wgpu/src/api/render_pipeline.rs:
```diff
@@ -121,6 +127,24 @@ pub struct RenderPipelineDescriptor<'a> {
     /// The layout of bind groups for this pipeline.
+    ///
+    /// If this is set, then `wgpu` will validate that the layout matches what the shader module(s)
+    /// expect, otherwise [`Device::create_render_pipeline`] will panic.
```
as above:
```suggestion
    /// expect, otherwise [`Device::create_render_pipeline`] will cause a validation error.
```

Wumpf reviewed (changes requested):
very nice thank you!
There's some fine print about `panic` here. Tbh I'm not entirely sure how to rectify this best: mismatching layout _is_ a panic by default. But as a user it doesn't have to be and that's a fairly important distinction. The full story is a bit lengthy though and (see #3767) not super well documented
Imho the suggestion I wrote in the comment works, but is not great.

BGR360 committed:
Apply suggestions from code review

Co-authored-by: Andreas Reich <r_andreas2@web.de>

BGR360 committed:
Remove text about panics in get_bind_group_layout.

We shouldn't advertise the behavior until it is consistent between
webgpu and wgpu-core.

Wumpf reviewed (approved):
thanks again, very appreciated!
